realm/realm-library/src/main/cpp/util.cpp

```diff
@@ -43,195 +43,165 @@ jmethodID session_error_handler;
 
 void ThrowRealmFileException(JNIEnv* env, const std::string& message, realm::RealmFileException::Kind kind);
 
-void ConvertException(JNIEnv* env, const char *file, int line)
-{
-    ostringstream ss;
+jthrowable JThrowableNew(JNIEnv* env, const char* class_name, const char* message) {
+    jclass j_exception_class = env->FindClass(class_name);
+    if (j_exception_class == nullptr) {
+        return nullptr;
+    }
+    jmethodID constructor = env->GetMethodID(j_exception_class, "<init>", "(Ljava/lang/String;)V");
+    if (constructor == nullptr) {
+        env->DeleteLocalRef(j_exception_class);
+        return nullptr;
+    }
+    jstring j_str = env->NewStringUTF(message);
+    jobject exception = env->NewObject(j_exception_class, constructor, j_str);
+    env->DeleteLocalRef(j_exception_class);
+    return reinterpret_cast<jthrowable>(exception);
+}
+
+jthrowable JThrowableNew(JNIEnv* env, const char* class_name, const char* message, jbyte kind) {
+    jclass j_exception_class = env->FindClass(class_name);
+    if (j_exception_class == nullptr) {
+        return nullptr;
+    }
+    jmethodID constructor = env->GetMethodID(j_exception_class, "<init>", "(BLjava/lang/String;)V");
+    if (constructor == nullptr) {
+        env->DeleteLocalRef(j_exception_class);
+        return nullptr;
+    }
+    jstring j_str = env->NewStringUTF(message);
+    jobject exception = env->NewObject(j_exception_class, constructor, kind, j_str);
+    env->DeleteLocalRef(j_exception_class);
+    return reinterpret_cast<jthrowable>(exception);
+}
+
+void ConvertException(JNIEnv* env) {
+    const char out_of_memory_error[]             = "io/realm/internal/OutOfMemoryError";
+    const char illegal_state_exception[]         = "java/lang/IllegalStateException";
+    const char illegal_argument_exception[]      = "java/lang/IllegalArgumentException";
+    const char realm_error[]                     = "io/realm/exceptions/RealmError";
+    const char bad_version[]                     = "io/realm/internal/async/BadVersionException";
+    const char realm_file_exception[]            = "io/realm/exceptions/RealmFileException";
+    const char index_out_of_bound_exception[]    = "java/lang/ArrayIndexOutOfBoundsException";
+    const char unsupported_operation_exception[] = "java/lang/UnsupportedOperationException";
+    const char class_not_found_exception[]       = "java/lang/ClassNotFoundException";
+    const char runtime_exception[]               = "java/lang/RuntimeException";
+
+    jthrowable j_exception;
+
+    // we are going to "throw" a Java exception so it is safest to clear
+    // already occurred exceptions
+    env->ExceptionClear();
```

Comment:
If there is a Java exception before, we'd better either just return to throw the old one / or do nothing just wait for a jni crash. I'd like the just have a JNI crash since it is a bug from our side and it should not happen.

Hide the old java exception may mislead our debbuging a lot.